Unable to upgrade to 0.3.0

I am trying to upgrade to the new release but unable to get it work.

Our code: https://github.com/open-web3-stack/boka/blob/72e7e3ac3b1dd5200ba3601aad13db46e26cc395/Tools/Sources/Tools/OpenRPC.swift#L103

It is generating JSON schema dynamically using reflection. I am not sure how to get it working with the new strongly typed generic code. The schema is dynamically generated and there is no way to know the type at the compile time. Any suggestions?

That will help with some of the error I am getting, but not this one: https://github.com/open-web3-stack/boka/actions/runs/12135367555/job/33834277425#step:12:12441

```
/runner/work/boka/boka/Tools/Sources/Tools/OpenRPC.swift:105:17: error: type '() -> any JSONSchemaComponent' cannot conform to 'JSONSchemaComponent'
103 |         return JSONObject {
104 |             info.properties.map { field in
105 |                 JSONProperty(key: field.name) {
    |                 |- error: type '() -> any JSONSchemaComponent' cannot conform to 'JSONSchemaComponent'
    |                 `- note: only concrete types such as structs, enums and classes can conform to protocols
106 |                     getSchema(type: field.type)
107 |                 }
```

`getSchema` is returning `any JSONSchemaComponent` and seems like Swift refuse to accept it as a generic argument

code: https://github.com/open-web3-stack/boka/blob/ccea92c7e26e72083c029ad3e19bbd442dc44e2d/Tools/Sources/Tools/OpenRPC.swift#L106

--- a/Sources/JSONSchemaBuilder/Builders/JSONPropertySchemaBuilder.swift
+++ b/Sources/JSONSchemaBuilder/Builders/JSONPropertySchemaBuilder.swift
@@ -25,6 +25,12 @@ import JSONSchema
   public static func buildEither<TrueComponent, FalseComponent>(
     second component: FalseComponent
   ) -> JSONPropertyComponents.Conditional<TrueComponent, FalseComponent> { .second(component) }
+
+  public static func buildArray<Component: PropertyCollection>(
+    _ components: [Component]
+  ) -> JSONPropertyComponents.PropertyArray<Component> {
+    .init(components: components)
+  }
 }
 
 public protocol PropertyCollection: Sendable {

--- a/Sources/JSONSchemaBuilder/JSONPropertyComponent/Modifier/PropertyArray.swift
+++ b/Sources/JSONSchemaBuilder/JSONPropertyComponent/Modifier/PropertyArray.swift
@@ -1,0 +1,38 @@
+import JSONSchema
+
+extension JSONPropertyComponents {
+  public struct PropertyArray<Component: PropertyCollection>: PropertyCollection {
+    let components: [Component]
+
+    public var requiredKeys: [String] {
+      components.flatMap(\.requiredKeys)
+    }
+
+    public var schemaValue: [String: JSONValue] {
+      components.reduce(into: [:]) { result, component in
+        result.merge(component.schemaValue) { current, _ in current }
+      }
+    }
+
+    public func validate(_ dictionary: [String: JSONValue]) -> Parsed<[Component.Output], ParseIssue> {
+      var outputs: [Component.Output] = []
+      var issues: [ParseIssue] = []
+
+      for component in components {
+        let result = component.validate(dictionary)
+        switch result {
+        case .valid(let output):
+          outputs.append(output)
+        case .invalid(let issue):
+          issues.append(contentsOf: issue)
+        }
+      }
+
+      if issues.isEmpty {
+        return .valid(outputs)
+      } else {
+        return .invalid(issues)
+      }
+    }
+  }
+}

--- a/Sources/JSONSchemaBuilder/JSONPropertyComponent/Modifier/PropertyArray.swift
+++ b/Sources/JSONSchemaBuilder/JSONPropertyComponent/Modifier/PropertyArray.swift
@@ -14,7 +14,9 @@ extension JSONPropertyComponents {
       }
     }
 
-    public func validate(_ dictionary: [String: JSONValue]) -> Parsed<[Component.Output], ParseIssue> {
+    public func validate(
+      _ dictionary: [String: JSONValue]
+    ) -> Parsed<[Component.Output], ParseIssue> {
       var outputs: [Component.Output] = []
       var issues: [ParseIssue] = []
 
@@ -28,11 +30,10 @@ extension JSONPropertyComponents {
         }
       }
 
-      if issues.isEmpty {
-        return .valid(outputs)
-      } else {
+      guard issues.isEmpty else {
         return .invalid(issues)
       }
+      return .valid(outputs)
     }
   }
 }

--- a/Tests/JSONSchemaBuilderTests/JSONPropertyTests.swift
+++ b/Tests/JSONSchemaBuilderTests/JSONPropertyTests.swift
@@ -88,6 +88,20 @@ struct JSONPropertySchemaTests {
     }
   }
 
+  @Test func multipleArray() throws {
+    let props = ["prop0", "prop1", "prop2", "prop3"]
+    @JSONPropertySchemaBuilder var sample: some PropertyCollection {
+      for prop in props {
+        JSONProperty(key: prop, value: JSONString())
+      }
+    }
+
+    try #require(sample.schemaValue.values.count == 4)
+    for value in sample.schemaValue.values {
+      #expect(value.object?["type"] == "string")
+    }
+  }
+
   @Test(arguments: [true, false]) func optional(_ bool: Bool) {
     @JSONPropertySchemaBuilder var sample: some PropertyCollection {
       if bool {

--- a/Tests/JSONSchemaBuilderTests/JSONSchemaTests.swift
+++ b/Tests/JSONSchemaBuilderTests/JSONSchemaTests.swift
@@ -311,7 +311,7 @@ struct JSONAdvancedBuilderTests {
     #expect(sample.schemaValue == (bool ? ["type": "number", "maximum": 100] : ["type": "number"]))
   }
 
-  func array(_ bool: Bool) {
+  @Test func array() {
     let properties = ["foo", "bar", "baz"]
 
     @JSONSchemaBuilder var sample: some JSONSchemaComponent {
@@ -323,7 +323,17 @@ struct JSONAdvancedBuilderTests {
         }
       }
     }
+    print("\(sample.schemaValue)")
 
-    #expect(sample.schemaValue == (bool ? ["type": "number", "maximum": 100] : ["type": "number"]))
+    let expected: [String: JSONValue] = [
+      "type": "object",
+      "properties": [
+        "foo": ["type": "string"],
+        "bar": ["type": "string"],
+        "baz": ["type": "string"],
+      ]
+    ]
+
+    #expect(sample.schemaValue == expected)
   }
 }

--- a/Tests/JSONSchemaBuilderTests/JSONSchemaTests.swift
+++ b/Tests/JSONSchemaBuilderTests/JSONSchemaTests.swift
@@ -310,4 +310,20 @@ struct JSONAdvancedBuilderTests {
 
     #expect(sample.schemaValue == (bool ? ["type": "number", "maximum": 100] : ["type": "number"]))
   }
+
+  func array(_ bool: Bool) {
+    let properties = ["foo", "bar", "baz"]
+
+    @JSONSchemaBuilder var sample: some JSONSchemaComponent {
+      JSONObject {
+        for property in properties {
+          JSONProperty(key: property) {
+            JSONString()
+          }
+        }
+      }
+    }
+
+    #expect(sample.schemaValue == (bool ? ["type": "number", "maximum": 100] : ["type": "number"]))
+  }
 }

--- a/Tests/JSONSchemaBuilderTests/JSONSchemaTests.swift
+++ b/Tests/JSONSchemaBuilderTests/JSONSchemaTests.swift
@@ -331,7 +331,7 @@ struct JSONAdvancedBuilderTests {
         "foo": ["type": "string"],
         "bar": ["type": "string"],
         "baz": ["type": "string"],
-      ]
+      ],
     ]
 
     #expect(sample.schemaValue == expected)

